// @genezio/auth/lib/models/UserTypes.js
"use strict";
/**
 * This is an auto generated code. This code should not be modified since the file can be overwritten
 * if new genezio commands are executed.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.ErrorCode = void 0;
var ErrorCode;
(function (ErrorCode) {
    ErrorCode[ErrorCode["CONNECTION_DATABASE_ERROR"] = 0] = "CONNECTION_DATABASE_ERROR";
    ErrorCode[ErrorCode["MISSING_EMAIL_PARAMETER"] = 1] = "MISSING_EMAIL_PARAMETER";
    ErrorCode[ErrorCode["INCORRECT_EMAIL_OR_PASSWORD"] = 2] = "INCORRECT_EMAIL_OR_PASSWORD";
    ErrorCode[ErrorCode["EMAIL_ALREADY_EXISTS"] = 3] = "EMAIL_ALREADY_EXISTS";
    ErrorCode[ErrorCode["EMAIL_NOT_FOUND"] = 4] = "EMAIL_NOT_FOUND";
    ErrorCode[ErrorCode["INVALID_TOKEN"] = 5] = "INVALID_TOKEN";
    ErrorCode[ErrorCode["MISSING_PASSWORD_PARAMETER"] = 6] = "MISSING_PASSWORD_PARAMETER";
    ErrorCode[ErrorCode["PASSWORD_TOO_SHORT"] = 7] = "PASSWORD_TOO_SHORT";
    ErrorCode[ErrorCode["PASSWORD_CONTAINS_ONLY_NUMBER"] = 8] = "PASSWORD_CONTAINS_ONLY_NUMBER";
    ErrorCode[ErrorCode["PASSWORD_MUST_HAVE_AT_LEAST_ONE_SPECIAL_CHARACTER"] = 9] = "PASSWORD_MUST_HAVE_AT_LEAST_ONE_SPECIAL_CHARACTER";
    ErrorCode[ErrorCode["INTERNAL_PASSWORD_NOT_HASHED_ERROR"] = 10] = "INTERNAL_PASSWORD_NOT_HASHED_ERROR";
    ErrorCode[ErrorCode["INTERNAL_CREATE_USER_ERROR"] = 11] = "INTERNAL_CREATE_USER_ERROR";
    ErrorCode[ErrorCode["INTERNAL_CREATE_SESSION_ERROR"] = 12] = "INTERNAL_CREATE_SESSION_ERROR";
    ErrorCode[ErrorCode["INTERNAL_UPDATE_USER_ERROR"] = 13] = "INTERNAL_UPDATE_USER_ERROR";
    ErrorCode[ErrorCode["INTERNAL_UPDATE_SESSION_ERROR"] = 14] = "INTERNAL_UPDATE_SESSION_ERROR";
    ErrorCode[ErrorCode["INTERNAL_DELETE_SESSION_ERROR"] = 15] = "INTERNAL_DELETE_SESSION_ERROR";
    ErrorCode[ErrorCode["INTERNAL_DELETE_USER_ERROR"] = 16] = "INTERNAL_DELETE_USER_ERROR";
    ErrorCode[ErrorCode["INTERNAL_FIND_USER_ERROR"] = 17] = "INTERNAL_FIND_USER_ERROR";
    ErrorCode[ErrorCode["INTERNAL_FIND_SESSION_ERROR"] = 18] = "INTERNAL_FIND_SESSION_ERROR";
    ErrorCode[ErrorCode["INTERNAL_INCORRECT_CONFIGRATION"] = 19] = "INTERNAL_INCORRECT_CONFIGRATION";
    ErrorCode[ErrorCode["MAIL_NOT_SENT"] = 20] = "MAIL_NOT_SENT";
    ErrorCode[ErrorCode["PASSWORD_MUST_HAVE_AT_LEAST_ONE_UPPERCASE_LETTER"] = 21] = "PASSWORD_MUST_HAVE_AT_LEAST_ONE_UPPERCASE_LETTER";
    ErrorCode[ErrorCode["EMAIL_NOT_VERIFIED"] = 22] = "EMAIL_NOT_VERIFIED";
    ErrorCode[ErrorCode["USER_NOT_FOUND"] = 23] = "USER_NOT_FOUND";
})(ErrorCode || (exports.ErrorCode = ErrorCode = {}));
;
